tensornetwork/network.py

```diff
@@ -913,6 +914,32 @@ def split_node_full_svd(
     self.nodes_set.remove(node)
     return left_node, singular_values_node, right_node, trun_vals
 
+  def remove_node(self, node: network_components.Node
+    ) -> Dict[Union[int, Text], network_components.Edge]:
+    """Remove a node from the network.
+
+    Args:
+      node: The node to be removed.
+
+    Returns:
+      A dictonary mapping name/index to the new broken edge.
+
+    Raises:
+      ValueError: If the node isn't in the network.
+    """
+    if node not in self:
+      raise ValueError("Node '{}' is not in the network.".format(node))
+    broken_edges = {}
+    for i, name in enumerate(node.axis_names):
+      if not node[i].is_dangling() and not node[i].is_trace():
+        edge1, edge2 = self.disconnect(node[i])
+        new_broken_edge = edge1 if edge1.node1 is not node else edge2
+        broken_edges[i] = new_broken_edge
+        broken_edges[name] = new_broken_edge
+    self.nodes_set.remove(node)
+    return broken_edges
```

Comment:
An alternative would be to return two dictionaries: one with broken edged by axis number and one with broken edges by axis name. I think the latter API is more convenient since in the current one the caller iterating over the dictionary must check key type if they're only interested in numbers/names. Also, they may accidentally double-process an edge.

Reply:
That's a really good point. However, that means that one of the dictionaries will be ignored most of the time, so we'll be seeing a lot of code that looks like `broke_edges, _ = net.remove_node(node)`

This isn't really that bad, and would certainly be preferable to checking the types when looping over the dictionary. Give me a day to think about it some more. If I can't figure out a better solution we'll do the two dictionaries.

Reply:
@amilsted @mganahl I would love to have your opinions on this.

Reply:
Tricky! Although it seems kind of ugly to return two dictionaries, but I agree that having two instances of each edge in the dict is asking for trouble. I also agree that it's not a big deal if one of the dictionaries is ignored, so I guess I think it would indeed be better to return two.

Reply:
I suppose an alternative would be to have an argument determine whether you get names or indices as keys, e.g. `return_axes_by_name=True`.

Reply:
I like the solution with an extra argument, but it hides a minor issue.

Making the argument a boolean spells some trouble with the type-checker down the road, since mypy won't be able to tell the return type of a given invocation.

We could solve this problem by defining two special values for the extra argument and using [overloads](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading):

```python
BY_NUMBER=-1
BY_NAME=""

@overload
def remove_node(self, node: Node, return_axes: int) -> Dict[int, Edge]: ...

@overload
def remove_node(self, node: Node, return_axes: str) -> Dict[str, Edge]: ...

def remove_node(self, node: Node, return_axes: Union[int, str]=BY_NAME) -> Union[Dict[int, Edge], Dict[str, Edge]]:
    if node...
    ...
    if return_axes is BY_NAME:
        return broken_edges_by_name
    return broken_edges_by_number

```

This way mypy will know that the key in the returned dictionary has the same type as the value passed in as `return_axes` argument (which is supposed to be either `BY_NUMBER` or BY_NAME which are an `int` and a `str` respectively).

Not sure whether default values play nicely with overloads, though.

Reply:
Yikes, I don't like that at all. It's pretty confusing and requires even more characters than `, _`. Alos, where would we store `BY_NAME`? Would it become a constant accessible via `tensor_network`? I don't wanna add a constant for this single method. 

Let's just return two dictionaries. Adding a `, _` really isnt that bad.